src/wallet-redux.js

```diff
@@ -22,137 +25,229 @@ export const WalletType = {
+        Object.assign({}, map.get(id), update),
+    )
+}
+
+function sanitizeActiveWalletId(state) {
+    const activeWalletIsEmptyLegacy = state.activeWalletId === LEGACY
+        && ![...state.accounts.values()].some(account => account.isLegacy);
+    const noActiveWallet = state.activeWalletId !== LEGACY && !state.wallets.get(state.activeWalletId);
+
+    if (activeWalletIsEmptyLegacy || noActiveWallet) {
+        // Determine wallet with most accounts and set it as new active wallet
+        const walletIdsWithCount = [...state.accounts.values()]
+            .map(account => account.isLegacy ? LEGACY : account.walletId)
+            .reduce((acc, id) => {
+                acc[id] = acc[id] || 0;
+                acc[id]++;
+                return acc;
+            }, {});
+
+        let highestCount = 0;
+        for (const id of Object.keys(walletIdsWithCount)) {
+            const count = walletIdsWithCount[id];
+            if (count > highestCount) {
+                state.activeWalletId = id;
+                highestCount = count;
+            }
+        }
+    }
+
+    return state;
+}
+
+/* Reducer: State machine where changes happen */
+
 export function reducer(state, action) {
+    // Dynamic helper function
+    function updateState(updatedProperties) {
+        const updatedState = Object.assign({}, state, updatedProperties);
+        return sanitizeActiveWalletId(updatedState);
+    }
+
     if (state === undefined) {
         return initialState;
     }
 
     switch (action.type) {
+        case TypeKeys.ADD_ACCOUNT:
+            const oldEntry = state.accounts.get(action.account.address);
+
+            // Keep balance information we might have
+            // TODO: Check if this is still used actually
+            return updateState({
+                accounts: new Map(state.accounts)
+                    .set(action.account.address, Object.assign({}, action.account, {
+                        balance: oldEntry ? oldEntry.balance : undefined
+                    }))
+            });
+
         case TypeKeys.LOGIN:
-            return Object.assign({}, state, {
-                hasContent: true,
-                entries: new Map(state.entries)
-                    .set(action.wallet.id, Object.assign({}, action.wallet)),
+            return updateState({
+                wallets: new Map(state.wallets).set(action.wallet.id, action.wallet),
                 activeWalletId: action.wallet.id,
             });
 
-        case TypeKeys.SWITCH:
-            return Object.assign({}, state, {
-                activeWalletId: action.walletId,
+        case TypeKeys.LOGOUT: {
+            // Remove wallet
+            const wallets = new Map(state.wallets);
+            wallets.delete(action.walletId);
+
+            // Remove wallet's accounts
+            const accounts = new Map(state.accounts);
+            for (const account of [...state.accounts.values()]) {
+                if (account.walletId === action.walletId) {
```

Comment:
This also finds contracts that belong to the wallet, so this works as expected (might add a comment that it also removes contracts by their walletId, as I had to search the code to validate that assumption).